### PR TITLE
[Serialization] Support loading template specializations lazily

### PR DESCRIPTION
**Pull request by ChuanqiXu9**
Reland https://github.com/llvm/llvm-project/pull/83237

---

(Original comments)

Currently all the specializations of a template (including instantiation, specialization and partial specializations)  will be loaded at once if we want to instantiate another instance for the template, or find instantiation for the template, or just want to complete the redecl chain.

This means basically we need to load every specializations for the template once the template declaration got loaded. This is bad since when we load a specialization, we need to load all of its template arguments. Then we have to deserialize a lot of unnecessary declarations.

For example,

```
// M.cppm
export module M;
export template <class T>
class A {};

export class ShouldNotBeLoaded {};

export class Temp {
   A<ShouldNotBeLoaded> AS;
};

// use.cpp
import M;
A<int> a;
```

We have a specialization ` A<ShouldNotBeLoaded>` in `M.cppm` and we instantiate the template `A` in `use.cpp`. Then we will deserialize `ShouldNotBeLoaded` surprisingly when compiling `use.cpp`. And this patch tries to avoid that.

Given that the templates are heavily used in C++, this is a pain point for the performance.

This patch adds MultiOnDiskHashTable for specializations in the ASTReader. Then we will only deserialize the specializations with the same template arguments. We made that by using ODRHash for the template arguments as the key of the hash table.

To review this patch, I think `ASTReaderDecl::AddLazySpecializations` may be a good entry point.

### CODE REVIEW

**Comment by h-vetinari** [comment]
I think the title of the commit ("Support load lazy specialization lazily") is hard to parse and somewhat redundant. Perhaps the following would be better (IIUC): "Support loading template specializations lazily"

Also "We should a specialization ` A<ShouldNotBeLoaded>` in `M.cppm` [...]" in the body of the commit message is hard to parse. Perhaps what was meant was "We have a specialization [...]"?

**Comment by ChuanqiXu9** [comment]
> I think the title of the commit ("Support load lazy specialization lazily") is hard to parse and somewhat redundant. Perhaps the following would be better (IIUC): "Support loading template specializations lazily"
> 
> Also "We should a specialization ` A<ShouldNotBeLoaded>` in `M.cppm` [...]" in the body of the commit message is hard to parse. Perhaps what was meant was "We have a specialization [...]"?

Nice catch. Thanks.

**Review by vgvassilev — APPROVED** [review]
Not sure if relands require reviews but lgtm!

**Comment by ChuanqiXu9** [comment]
Thanks. Given the change in lldb is trivial, I'd like to land it to see what happens.

**Comment by vgvassilev** [comment]
@ChuanqiXu9, can we insert a fix for windows?

**Comment by hahnjo** [comment]
I believe https://github.com/llvm/llvm-project/commit/30ea0f0ce476bf4c12684a9a514af2ca660bbe44 already addresses the cast. The bots are just slow...

**Comment by ilya-biryukov** [comment]
@ChuanqiXu9 the added test fails under ASAN that reports memory leaks:

```
$ cmake -G Ninja -DLLVM_USE_SANITIZER=Address -DCMAKE_CXX_COMPILER=clang -DCMAKE_C_COMPILER=clang ../
$ ninja SerializationTests
$ ./tools/clang/unittests/Serialization/SerializationTests
```

produces
```
==4100640==ERROR: LeakSanitizer: detected memory leaks

Indirect leak of 32776 byte(s) in 1 object(s) allocated from:
    #0 0x559f90ba330c in calloc (/usr/local/google/home/ibiryukov/code/llvm-project/build-asan/tools/clang/unittests/Serialization/SerializationTests+0x174d30c) (BuildId: 4baf7bfc387b0be3)
    #1 0x559f910585eb in safe_calloc /usr/local/google/home/ibiryukov/code/llvm-project/llvm/include/llvm/Support/MemAlloc.h:40:18
    #2 0x559f910585eb in AllocateBuckets /usr/local/google/home/ibiryukov/code/llvm-project/llvm/lib/Support/FoldingSet.cpp:173:40
    #3 0x559f910585eb in llvm::FoldingSetBase::FoldingSetBase(unsigned int) /usr/local/google/home/ibiryukov/code/llvm-project/llvm/lib/Support/FoldingSet.cpp:187:13
    #4 0x559f90f34503 in FoldingSetImpl /usr/local/google/home/ibiryukov/code/llvm-project/llvm/include/llvm/ADT/FoldingSet.h:454:9
    #5 0x559f90f34503 in ContextualFoldingSet /usr/local/google/home/ibiryukov/code/llvm-project/llvm/include/llvm/ADT/FoldingSet.h:636:9
    #6 0x559f90f34503 in clang::ASTContext::ASTContext(clang::LangOptions&, clang::SourceManager&, clang::IdentifierTable&, clang::SelectorTable&, clang::Builtin::Context&, clang::TranslationUnitKind) /usr/local/google/home/ibiryukov/code/llvm-project/clang/lib/AST/ASTContext.cpp:920:7
    #7 0x559f92667204 in clang::CompilerInstance::createASTContext() /usr/local/google/home/ibiryukov/code/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:553:23
    #8 0x559f928ad79a in clang::FrontendAction::BeginSourceFile(clang::CompilerInstance&, clang::FrontendInputFile const&) /usr/local/google/home/ibiryukov/code/llvm-project/clang/lib/Frontend/FrontendAction.cpp:948:10
    #9 0x559f9266fb7d in clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /usr/local/google/home/ibiryukov/code/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:1061:13
    #10 0x559f90c13901 in (anonymous namespace)::LoadSpecLazilyTest::GenerateModuleInterface[abi:cxx11](llvm::StringRef, llvm::StringRef) /usr/local/google/home/ibiryukov/code/llvm-project/clang/unittests/Serialization/LoadSpecLazilyTest.cpp:86:5
    #11 0x559f90c18b8b in (anonymous namespace)::LoadSpecLazilyTest_ChainedTest2_Test::TestBody() /usr/local/google/home/ibiryukov/code/llvm-project/clang/unittests/Serialization/LoadSpecLazilyTest.cpp:227:3
    #12 0x559f90e46395 in HandleExceptionsInMethodIfSupported<testing::Test, void> /usr/local/google/home/ibiryukov/code/llvm-project/third-party/unittest/googletest/src/gtest.cc
    #13 0x559f90e46395 in testing::Test::Run() /usr/local/google/home/ibiryukov/code/llvm-project/third-party/unittest/googletest/src/gtest.cc:2687:5
    #14 0x559f90e48ae2 in testing::TestInfo::Run() /usr/local/google/home/ibiryukov/code/llvm-project/third-party/unittest/googletest/src/gtest.cc:2836:11
    #15 0x559f90e4ad74 in testing::TestSuite::Run() /usr/local/google/home/ibiryukov/code/llvm-project/third-party/unittest/googletest/src/gtest.cc:3015:30
    #16 0x559f90e82ee8 in testing::internal::UnitTestImpl::RunAllTests() /usr/local/google/home/ibiryukov/code/llvm-project/third-party/unittest/googletest/src/gtest.cc:5920:44
    #17 0x559f90e811c7 in HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> /usr/local/google/home/ibiryukov/code/llvm-project/third-party/unittest/googletest/src/gtest.cc
    #18 0x559f90e811c7 in testing::UnitTest::Run() /usr/local/google/home/ibiryukov/code/llvm-project/third-party/unittest/googletest/src/gtest.cc:5484:10
    #19 0x559f90e08bfa in RUN_ALL_TESTS /usr/local/google/home/ibiryukov/code/llvm-project/third-party/unittest/googletest/include/gtest/gtest.h:2317:73
    #20 0x559f90e08bfa in main /usr/local/google/home/ibiryukov/code/llvm-project/third-party/unittest/UnitTestMain/TestMain.cpp:55:10
    #21 0x7f0546213c89 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
...
SUMMARY: AddressSanitizer: 488922 byte(s) leaked in 529 allocation(s).
```

Maybe revert the change or is there a simple fix forward?

**Comment by ilya-biryukov** [comment]
Ah, it's `DisableFree` again. I'll send a fix.

**Comment by ilya-biryukov** [comment]
Fixed by 7f4312015291a32d811a0f37e24b4d9736c524f7.

File: clang/include/clang/Serialization/ASTBitCodes.h [hunk 1]
```diff
@@ -1502,6 +1509,12 @@ enum DeclCode {
   /// An ImplicitConceptSpecializationDecl record.
   DECL_IMPLICIT_CONCEPT_SPECIALIZATION,
 
+  // A decls specilization record.
```
Comment (erichkeane) [inline clang/include/clang/Serialization/ASTBitCodes.h:1512]: Note `specialization` is misspelled in this comment and the one below.

File: clang/include/clang/Serialization/ASTBitCodes.h [hunk 2]
```diff
@@ -1502,6 +1509,12 @@ enum DeclCode {
...
+  DECL_SPECIALIZATIONS,
+
+  // A decls specilization record.
+  DECL_PARTIAL_SPECIALIZATIONS,
+
```
Comment (erichkeane) [inline clang/include/clang/Serialization/ASTBitCodes.h:1517]: The `DECL_LAST` also needs to be updated her.e

Comment (ChuanqiXu9) [reply clang/include/clang/Serialization/ASTBitCodes.h:1517]: Thanks for catching this! Done in https://github.com/llvm/llvm-project/commit/366daddfad9aa38ebb7d40055cf65f4ecb7dd6f9

File: clang/include/clang/Serialization/ASTBitCodes.h (hunk 1 above)
Comment (ChuanqiXu9) [reply clang/include/clang/Serialization/ASTBitCodes.h:1512]: Done

**Comment by michael-jabbour-sonarsource** [comment]
Hello,

We have encountered crashes after this patch, with stack traces that look like this:

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
  ...
 #0 0x0000740f12c266f8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) 
 #1 0x0000740f12c23de5 llvm::sys::RunSignalHandlers()
 #2 0x0000740f12c274b1 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x0000740f12242520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x0000740f10a04134 clang::Redeclarable<clang::TagDecl>::DeclLink::getPrevious(clang::TagDecl const*) const APValue.cpp:0:0
 #5 0x0000740f10a99531 clang::CXXRecordDecl::getMostRecentNonInjectedDecl() ASTDumper.cpp:0:0
 #6 0x0000740f10d71be7 clang::RecursiveASTVisitor<clang::ParentMapContext::ParentMap::ASTVisitor>::TraverseTemplateInstantiations(clang::ClassTemplateDecl*)
 #7 0x0000740f10d4cb37 clang::RecursiveASTVisitor<clang::ParentMapContext::ParentMap::ASTVisitor>::TraverseClassTemplateDecl(clang::ClassTemplateDecl*)
 #8 0x0000740f10d46926 clang::RecursiveASTVisitor<clang::ParentMapContext::ParentMap::ASTVisitor>::TraverseDecl(clang::Decl*)
 #9 0x0000740f10d6ac21 clang::RecursiveASTVisitor<clang::ParentMapContext::ParentMap::ASTVisitor>::TraverseDeclContextHelper(clang::DeclContext*)
#10 0x0000740f10d4d0c4 clang::RecursiveASTVisitor<clang::ParentMapContext::ParentMap::ASTVisitor>::TraverseNamespaceDecl(clang::NamespaceDecl*)
#11 0x0000740f10d46b9c clang::RecursiveASTVisitor<clang::ParentMapContext::ParentMap::ASTVisitor>::TraverseDecl(clang::Decl*)
#12 0x0000740f10d6ac21 clang::RecursiveASTVisitor<clang::ParentMapContext::ParentMap::ASTVisitor>::TraverseDeclContextHelper(clang::DeclContext*)
#13 0x0000740f10d46f87 clang::RecursiveASTVisitor<clang::ParentMapContext::ParentMap::ASTVisitor>::TraverseTranslationUnitDecl(clang::TranslationUnitDecl*)
#14 0x0000740f10d468a0 clang::RecursiveASTVisitor<clang::ParentMapContext::ParentMap::ASTVisitor>::TraverseDecl(clang::Decl*)
#15 0x0000740f10d45d38 clang::ParentMapContext::ParentMap::ASTVisitor::TraverseDecl(clang::Decl*) ParentMapContext.cpp:0:0
#16 0x0000740f10d42801 clang::ParentMapContext::ParentMap::ParentMap(clang::ASTContext&)
#17 0x0000740f10d42876 clang::ParentMapContext::getParents(clang::DynTypedNode const&)
#18 0x0000740f1750739b (anonymous namespace)::ScalarExprEmitter::EmitScalarPrePostIncDec(clang::UnaryOperator const*, clang::CodeGen::LValue, bool, bool) CGExprScalar.cpp:0:0
#19 0x0000740f175132be (anonymous namespace)::ScalarExprEmitter::VisitUnaryPostDec(clang::UnaryOperator const*) CGExprScalar.cpp:0:0
#20 0x0000740f175051eb (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) CGExprScalar.cpp:0:0
#21 0x0000740f17521bdd (anonymous namespace)::ScalarExprEmitter::VisitCastExpr(clang::CastExpr*) CGExprScalar.cpp:0:0
#22 0x0000740f175050dd clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool)
#23 0x0000740f1748fe28 clang::CodeGen::CodeGenFunction::EvaluateExprAsBool(clang::Expr const*)
#24 0x0000740f176722d3 clang::CodeGen::CodeGenFunction::EmitWhileStmt(clang::WhileStmt const&, llvm::ArrayRef<clang::Attr const*>)
#25 0x0000740f17670edf clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>)
#26 0x0000740f1767ed60 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot)
#27 0x0000740f176f765e clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) 
#28 0x0000740f17721d6e clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*)
#29 0x0000740f1771950e clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) 
#30 0x0000740f1771e832 clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) 
#31 0x0000740f17718011 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) 
#32 0x0000740f177eb7bc (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) ModuleBuilder.cpp:0:0
#33 0x0000740f176e2726 clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) 
#34 0x0000740f11af6f8a clang::ParseAST(clang::Sema&, bool, bool) 
#35 0x0000740f155ee356 clang::FrontendAction::Execute() 
#36 0x0000740f1555060d clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) 
#37 0x0000740f18a75ae7 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) 
#38 0x000063e90df5329f cc1_main(llvm::ArrayRef<char const*>, char const*, void*) 
#39 0x000063e90df4f62d ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#40 0x000063e90df4e669 clang_main(int, char**, llvm::ToolContext const&) 
#41 0x000063e90df601a7 main 
#42 0x0000740f12229d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#43 0x0000740f12229e40 call_init ./csu/../csu/libc-start.c:128:20
#44 0x0000740f12229e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#45 0x000063e90df4d0c5 _start 
```
I have created a fix and a test case in #150430. Any feedback is appreciated.

CC: @ChuanqiXu9, @erichkeane, @vgvassilev
